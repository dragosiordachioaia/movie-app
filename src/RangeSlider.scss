@import "variables";

.range-slider {
  position: relative;
  width: 100%;
  margin-bottom: 20px;

  .axis {
    width: 100%;
    height: 2px;
    background-color: $color-primary;
    margin-bottom: 10px;
  }

  .cursor {
    position: absolute;
    transform: translate(-50%, -50%);
    left: 0;
    top: 0;
    width: 0.8em;
    height: 0.8em;
    background-color: $color-secondary-alt;
    cursor: pointer;
    border-radius: 50%;
  }

  .min {
    float: left;
  }

  .max {
    float: right;
  }
}
