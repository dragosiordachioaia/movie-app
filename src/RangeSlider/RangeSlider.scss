@import "../variables";

.range-slider {
  position: relative;
  width: 100%;
  margin-bottom: 1.5em;
  margin-top: 2em;

  @media (min-width: $SM) {
    margin-top: 2.5em;
  }

  .axis {
    width: 100%;
    height: 2px;
    background-color: $color-primary;
    margin-bottom: 10px;
  }

  .axis-highlight {
    height: 2px;
    position: absolute;
    background-color: $color-secondary-alt;
    top: 0;
    pointer-events: none;
    user-select: none;
    transition: width 250ms;
  }

  .cursor {
    position: absolute;
    transform: translate(-50%, -50%);
    left: 0;
    top: 0;
    width: 0.8em;
    height: 0.8em;
    background-color: $color-secondary-alt;
    cursor: pointer;
    border-radius: 50%;
    transition: left 250ms;
  }

  .tooltip {
    user-select: none;
    pointer-events: none;
    // background-color: #fff;
    color: #fff;
    display: inline-block;
    width: 2em;
    text-align: center;
    padding-top: 0.2em;
    padding-bottom: 0.2em;
    transform: translate(-50%, -50%);
    top: -1.5em;
    position: absolute;
    transition: left 250ms;
  }

  .min,
  .max,
  .tooltip {
    user-select: none;
    font-weight: 700;
    font-size: 0.8em;
  }

  .min {
    float: left;
    margin-top: -0.4em;
  }

  .max {
    float: right;
    margin-top: -0.4em;
  }
}
