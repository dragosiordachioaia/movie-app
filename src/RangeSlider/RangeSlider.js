import React, { Component } from "react";

import "./RangeSlider.scss";

import { roundToStep } from "utils";

export default class RangeSlider extends Component {
  constructor(props) {
    super(props);
    this.state = {
      mouseDown: false,
      touchStart: false,
    };
    this.onMouseMove = this.onMouseMove.bind(this);
    this.onTouchMove = this.onTouchMove.bind(this);
    this.onMouseUp = this.onMouseUp.bind(this);
    this.onTouchEnd = this.onTouchEnd.bind(this);
    this.callOnChange = this.callOnChange.bind(this);

    this.lastMousePos = 0;
    this.axisRef = React.createRef();
  }

  componentDidMount() {
    window.addEventListener("mouseup", this.onMouseUp);
    window.addEventListener("touchend", this.onTouchEnd);
    window.addEventListener("mousemove", this.onMouseMove);
    window.addEventListener("touchmove", this.onTouchMove);
  }

  onMouseUp(e) {
    this.setState({ mouseDown: false });
  }
  onTouchEnd(e) {
    this.setState({ touchStart: false });
  }

  onMouseMove(e) {
    if (!this.state.mouseDown) {
      return;
    }
    let axisElement = this.axisRef.current;
    let axisWidth = axisElement.offsetWidth;
    let axisLeft = axisElement.getBoundingClientRect().left;
    let position = Math.round(e.screenX - axisLeft);
    position = Math.min(axisWidth, Math.max(position, 0));
    let percentage = position / axisWidth;
    this.callOnChange(percentage);
  }

  callOnChange(percentage) {
    if (this.props.onChange && typeof this.props.onChange === "function") {
      let minMaxDifference = this.props.max - this.props.min;
      let realAmount = roundToStep(
        percentage * minMaxDifference,
        this.props.step
      );
      this.props.onChange(realAmount);
    }
  }

  onTouchMove(e) {
    if (!this.state.touchStart) {
      return;
    }
    let axisElement = this.axisRef.current;
    let axisWidth = axisElement.offsetWidth;
    let axisLeft = axisElement.getBoundingClientRect().left;
    let position = Math.round(e.changedTouches[0].pageX - axisLeft);
    position = Math.min(axisWidth, Math.max(position, 0));
    let percentage = position / axisWidth;
    this.callOnChange(percentage);
  }

  render() {
    let cursorPosition =
      (this.props.value / (this.props.max - this.props.min)) * 100;
    return (
      <div className="range-slider">
        <div
          className="cursor"
          style={{ left: `${cursorPosition}%` }}
          onMouseDown={e => this.setState({ mouseDown: true })}
          onTouchStart={e => this.setState({ touchStart: true })}
        />
        <div className="tooltip" style={{ left: `${cursorPosition}%` }}>
          {this.props.value}
        </div>
        <div className="axis" ref={this.axisRef} />
        <div
          className="axis-highlight"
          style={{ width: `${cursorPosition}%` }}
        />
        <label className="min">{this.props.min}</label>
        <label className="max">{this.props.max}</label>
      </div>
    );
  }
}
