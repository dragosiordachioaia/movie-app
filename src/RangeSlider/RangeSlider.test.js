import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import RangeSlider from "./RangeSlider";

// jest.mock("RangeSlider/RangeSlider", () => "RangeSlider");

let component;
let instance;

beforeEach(() => {
  component = renderer.create(<RangeSlider />);
  instance = component.getInstance();
});

it("renders correctly", () => {
  instance.props = {
    value: 3,
    min: 0,
    max: 10,
  };
  const result = instance.render();
  expect(result).toMatchSnapshot();
});

it("does not mouse record movement if mouseDown is false", () => {
  let mockOnChange = jest.fn();
  instance.state = {
    mouseDown: false,
  };
  instance.onMouseMove({});
  expect(mockOnChange).not.toBeCalled();
});

it("does not touch record movement if touchStart is false", () => {
  let mockOnChange = jest.fn();
  instance.state = {
    touchStart: false,
  };
  instance.onTouchMove({});
  expect(mockOnChange).not.toBeCalled();
});

it("correctly records mouse movement if mouseDown is true", () => {
  let mockOnChange = jest.fn();
  instance.state = {
    mouseDown: true,
  };
  instance.axisRef = {
    current: {
      offsetWidth: 100,
      getBoundingClientRect: () => {
        return {
          left: 10,
        };
      },
    },
  };
  let mockCallOnChange = jest.fn();
  instance.callOnChange = mockCallOnChange;
  instance.onMouseMove({ screenX: 20 });
  expect(mockCallOnChange).toBeCalledWith(0.1);
});

it("correctly records touch movement if touchStart is true", () => {
  let mockOnChange = jest.fn();
  instance.state = {
    touchStart: true,
  };
  instance.axisRef = {
    current: {
      offsetWidth: 100,
      getBoundingClientRect: () => {
        return {
          left: 10,
        };
      },
    },
  };
  let mockCallOnChange = jest.fn();
  instance.callOnChange = mockCallOnChange;
  instance.onTouchMove({
    changedTouches: [{ pageX: 20 }],
  });
  expect(mockCallOnChange).toBeCalledWith(0.1);
});

it("correctly sends amount on change", () => {
  let mockOnChange = jest.fn();
  instance.props = {
    onChange: mockOnChange,
    min: 0,
    max: 10,
    step: 0.5,
  };
  instance.callOnChange(0.33);
  expect(mockOnChange).toBeCalledWith("3.5");
});
