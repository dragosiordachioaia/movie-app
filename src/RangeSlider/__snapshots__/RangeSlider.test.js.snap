// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="range-slider"
>
  <div
    className="cursor"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "left": "30%",
      }
    }
  />
  <div
    className="tooltip"
    style={
      Object {
        "left": "30%",
      }
    }
  >
    3
  </div>
  <div
    className="axis"
  />
  <div
    className="axis-highlight"
    style={
      Object {
        "width": "30%",
      }
    }
  />
  <label
    className="min"
  >
    0
  </label>
  <label
    className="max"
  >
    10
  </label>
</div>
`;
